clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java

```
@@ -181,36 +178,58 @@ public synchronized void unsubscribe() {
             if (!subscriptions.hasValidPosition(tp))
                 updateFetchPosition(tp);
 
-        // update the consumed offset
         final Map<TopicPartition, List<ConsumerRecord<K, V>>> results = new HashMap<>();
-
-        for (Map.Entry<TopicPartition, List<ConsumerRecord<K, V>>> entry : this.records.entrySet()) {
-            if (!subscriptions.isPaused(entry.getKey())) {
-                final List<ConsumerRecord<K, V>> recs = entry.getValue();
-                for (final ConsumerRecord<K, V> rec : recs) {
-                    if (beginningOffsets.get(entry.getKey()) != null && beginningOffsets.get(entry.getKey()) > subscriptions.position(entry.getKey())) {
-                        throw new OffsetOutOfRangeException(Collections.singletonMap(entry.getKey(), subscriptions.position(entry.getKey())));
-                    }
-
-                    if (assignment().contains(entry.getKey()) && rec.offset() >= subscriptions.position(entry.getKey())) {
-                        results.computeIfAbsent(entry.getKey(), partition -> new ArrayList<>()).add(rec);
-                        subscriptions.position(entry.getKey(), rec.offset() + 1);
-                    }
+        final Set<TopicPartition> partitionsToPoll = assignment().stream()
+                .filter(tp -> !subscriptions.isPaused(tp))
+                .collect(Collectors.toSet());
+        for (TopicPartition tp : partitionsToPoll) {
+            List<ConsumerRecord<K, V>> records;
+            try {
+                records = getLog(tp).fetch(subscriptions.position(tp));
+            } catch (OffsetOutOfRangeException e) {
+                if (subscriptions.hasDefaultOffsetResetPolicy()) {
```

In the `else` case, should we raise this exception?

---

This should already happen, since we call fetch again a few lines below. I can see how it isn't obvious though, so I'll add an `else` that rethrows.